lib/rules/must-tear-down-animations.js

```diff
@@ -1,28 +1,156 @@
 
 const create = (context) => {
 
+    // TODO(amy): express this in a way where names are scoped
+    // within a single component to handle multi-component cases.
+    // Using a flat object temporarily to get the simplest cases covered.
+
+    // Stores all animated values that have been set.
+    // When an animation is torn down, it is removed.
+    const activeAnimations = {};
+
     /**
      * Report teardown violations (animations that are set but not
      * torn down in componentWillUnmount)
      */
-    function reportMissingTeardowns(errorNodes) {
+    function reportMissingTeardowns() {
+        // Get list of all animated nodes that remain active (not torn down)
+        const activeNodes = Object.values(activeAnimations);
         // Iterate through all nodes failing the lint rule
         // and use context.report to surface error to user.
-        errorNodes.forEach(node => {
+        activeNodes.forEach(node => {
             context.report({
                 node: node,
-                message: 'Must tear down animation on unmount',
+                message: `Must tear down animations when component unmounts`,
             });
         });
     }
-    // List where we will accumulate nodes violating the rule.
-    const errorNodes = [];
+
+    /**
+     * Check state initialization. If state is set to an animated value
+     * directly we record the animation.
+     */
+    function checkAnimatedStateInitialization(node) {
+        const stateProperty = _getStateInitialization(node);
+        if (stateProperty) {
+            // If initializing state with an animation directly, record the animation
+            const expression = node.value.callee;
+            const isAnimation = astHelpers.isAnimationDeclaration(expression);
+            if (isAnimation) {
+                activeAnimations[stateProperty] = node.value;
+            }
+        }
+    }
+
+    /**
+     * Check setState calls. If state is set to an animated value directly,
+     * we record the animation.
+     */
+    function checkSetState(node) {
+        const newState = _getStateUpdate(node);
+        if (newState) {
+            newState.properties.forEach(p => {
+                const stateProperty = p.key.name;
+                const value = p.value;
+                // If setting state to an animation directly, record the animation
+                const isAnimation = astHelpers.isAnimationDeclaration(value.callee);
+                if (isAnimation) {
+                    activeAnimations[stateProperty] = value;
+                }
+            })
+        }
+    }
+
+    /**
+     * Detect if state is being set via setState (`this.setState({foo: "bar"})`).
+     * Returns the new state object.
+     */
+    function _getStateUpdate(node) {
+        if (node.callee && node.callee.property &&
+            node.callee.property.name === "setState") {
+            // We assume setState always takes a single argument.
```

Comment:
We have a few places that we use a second argument to `setState` (a function that's called when we finish setting state), but we should always have at least one argument!

(Probably fine to leave the comment as is, just adding thoughts around the assumption and 👍 )